src/template.rs

```diff
@@ -0,0 +1,137 @@
+use crate::Event;
+use bytes::Bytes;
+use lazy_static::lazy_static;
+use regex::bytes::{Captures, Regex};
+use string_cache::DefaultAtom as Atom;
+
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"\{\{(?P<key>[^\}]+)\}\}").unwrap();
+}
+
+#[derive(Debug, Clone)]
+pub struct Template {
+    src: Bytes,
+}
+
+impl From<&str> for Template {
+    fn from(src: &str) -> Template {
+        Template { src: src.into() }
+    }
+}
+
+impl Template {
+    pub fn render(&self, event: &Event) -> Result<Bytes, Vec<Atom>> {
+        let mut missing_fields = Vec::new();
+        let out = RE
+            .replace_all(self.src.as_ref(), |caps: &Captures| {
+                let key = caps
+                    .get(1)
+                    .and_then(|m| std::str::from_utf8(m.as_bytes()).ok())
+                    .map(|s| Atom::from(s))
+                    .expect("src should match regex and keys should be utf8");
```

Comment:
how do we uphold this invariant? by only letting it be constructed from a string/str?

Reply:
Yeah, non-optional capture group should always be present and we construct from a `str` so the key should always be valid utf-8. Annoying, but after playing around a bit I think we just stick with it.